File: PostgresRPQ/RPQEngine.py
import sys
import time
import psycopg2
from antlr3  import *
from antlr3.tree import *
from RegExpLexer import RegExpLexer
from RegExpParser import RegExpParser

class TreeWalker:
    
    def __init__(self, root, isFour):
        self.isFour = isFour
        self.root = root
    
    def postorder(self, tree):
        for child in tree.getChildren():
            self.postorder(child)
        subexpstr = ""
        for child in tree.getChildren():
            subexpstr += child.subexp.value + " "
        subexpstr += tree.text + " "
        tree.subexp = lambda: None
        setattr(tree.subexp, 'value', subexpstr.strip())
        print tree.text
        
    def doWalk(self, tree):
        self.initializeQuery()
        self.walk(tree)
        self.closeQuery()
        
    def walk(self, tree):
        for child in tree.getChildren():
            self.walk(child)
        if tree.getText() == "|":
            self.outputOr(tree)
        elif tree.getText() == ".":
            self.outputConc(tree)
        elif tree.getText() == "*":
            self.outputStar(tree)
        elif tree.getText() == "+":
            self.outputPlus(tree)
        elif tree.getText() == "-":
            self.outputMinus(tree)
        elif tree.getText() == "?":
            self.outputOptional(tree)
        else:
            pass


    def outputOr(self, node):
        if (self.isFour):
            orQuery = '''INSERT INTO g SELECT compstart, %s, compend, basestart, label2, baseend 
                         FROM g WHERE label1 = %s OR label1 = %s '''
        else:
            orQuery = '''INSERT INTO g SELECT compstart, %s, compend 
                         FROM g WHERE label1 = %s OR label1 = %s '''
        valueTuple = (node.subexp.value, node.getChild(0).subexp.value, node.getChild(1).subexp.value)
        self.cur.execute(orQuery, valueTuple)

    def outputConc(self, node):
        if (self.isFour):
            concQuery = '''INSERT INTO g
                SELECT g1.compstart, %s, g2.compend, g1.basestart, g1.label2, g1.baseend 
                FROM g as g1, g as g2 
                WHERE g1.label1 = %s and g1.compend = g2.compstart and g2.label1 = %s'''
            valueTuple = (node.subexp.value, node.getChild(0).subexp.value, node.getChild(1).subexp.value)
            self.cur.execute(concQuery, valueTuple)
            concQuery = '''INSERT INTO g
                SELECT g1.compstart, %s, g2.compend, g2.basestart, g2.label2, g2.baseend 
                FROM g as g1, g as g2 
                WHERE g1.label1 = %s and g1.compend = g2.compstart and g2.label1 = %s'''
            self.cur.execute(concQuery, valueTuple)
        else:
            concQuery = '''INSERT INTO g
                SELECT g1.compstart, %s, g2.compend 
                FROM g as g1, g as g2 
                WHERE g1.label1 = %s and g1.compend = g2.compstart and g2.label1 = %s'''
            valueTuple = (node.subexp.value, node.getChild(0).subexp.value, node.getChild(1).subexp.value)
            self.cur.execute(concQuery, valueTuple)

    def outputStar(self, node):
        if (self.isFour):
            tcQuery = '''select * from tcwithrecursive4ary(%s, %s)'''
            self.cur.execute(tcQuery, (node.getChild(0).subexp.value, ' *'))
            selfedgeQueryleft = '''INSERT INTO g
                select g.basestart, %s, g.basestart, g.basestart, %s, g.basestart from g where g.label1 = g.label2'''
            selfedgeQueryright = '''INSERT INTO g
                select g.baseend, %s, g.baseend, g.baseend, %s, g.baseend from g where g.label1 = g.label2'''
            self.cur.execute(selfedgeQueryleft, (node.subexp.value, node.subexp.value))
            self.cur.execute(selfedgeQueryright, (node.subexp.value, node.subexp.value))
        else:
            tcQuery = '''select * from tcwithrecursive(%s, %s)'''
            self.cur.execute(tcQuery, (node.getChild(0).subexp.value, ' *'))
            selfedgeQuery = '''INSERT INTO g
                select distinct g.compstart, %s, g.compstart from g
                union
                select distinct g.compend, %s, g.compend from g'''
            self.cur.execute(selfedgeQuery, (node.subexp.value, node.subexp.value))         

    def outputPlus(self, node):
        if (self.isFour):
            tcQuery = '''select * from tcwithrecursive4aryintersect(%s, %s)'''
            self.cur.execute(tcQuery, (node.getChild(0).subexp.value, ' +'))
        else:
            tcQuery = '''select * from tcwithrecursive(%s, %s)'''
            self.cur.execute(tcQuery, (node.getChild(0).subexp.value, ' +'))
        
    def outputMinus(self, node):
        valueTuple = (node.subexp.value, node.getChild(0).subexp.value)
        if (self.isFour):
            self.cur.execute(''' INSERT INTO g SELECT compend, %s, compstart, basestart, label2, baseend
                                 FROM g WHERE label1 = %s; ''', valueTuple)  
        else:
            self.cur.execute("INSERT INTO g SELECT compend, %s, compstart FROM g WHERE label1 = %s;", valueTuple)
    
    def outputOptional(self, node):
        if (self.isFour):
            selfedgeQueryleft = '''INSERT INTO g
                select g.basestart, %s, g.basestart, g.basestart, %s, g.basestart from g where g.label1 = g.label2'''
            selfedgeQueryright = '''INSERT INTO g
                select g.baseend, %s, g.baseend, g.baseend, %s, g.baseend from g where g.label1 = g.label2'''
            self.cur.execute(selfedgeQueryleft, (node.subexp.value, node.getChild(0).subexp.value))
            self.cur.execute(selfedgeQueryright, (node.subexp.value, node.getChild(0).subexp.value))
            optQuery = '''INSERT INTO g SELECT compstart, %s, compend, basestart, label2, baseend FROM g WHERE label1 = %s'''
            valueTuple = (node.subexp.value, node.getChild(0).subexp.value)
            self.cur.execute(optQuery, valueTuple)
        else:
            selfedgeQueryleft = '''INSERT INTO g select g.compstart, %s, g.compstart from g'''
            selfedgeQueryright = '''INSERT INTO g select g.compend, %s, g.compend from g'''
            self.cur.execute(selfedgeQueryleft, (node.subexp.value,))
            self.cur.execute(selfedgeQueryright, (node.subexp.value,))
            optQuery = '''INSERT INTO g SELECT compstart, %s, compend FROM g WHERE label1 = %s'''
            valueTuple = (node.subexp.value, node.getChild(0).subexp.value)
            self.cur.execute(optQuery, valueTuple)


    def initializeQuery(self):
        self.t0 = time.time()
        print("Query starting...")
        configFile = open(sys.argv[3])
        dbname = configFile.readline()
        username = configFile.readline()
        password = configFile.readline()
        configFile.close()
        connectionString = "host=localhost dbname=" + dbname + " user=" + username + " password=" + password
        self.conn = psycopg2.connect(connectionString)
        self.cur = self.conn.cursor()

        if (self.isFour):
            self.cur.execute( '''CREATE TABLE g (
                compstart character varying(100),
                label1 character varying(45),
                compend character varying(100),
                basestart character varying(100),
                label2 character varying(45),
                baseend character varying(100) );
                    ''' )
                    # CREATE INDEX gidx ON g (label1);
            
            insertionQuery = "INSERT INTO g select *, * from " + sys.argv[2]

        else:
            self.cur.execute('''CREATE TABLE g (
                compstart character varying(100),
                label1 character varying(45),
                compend character varying(100) );
                    ''' )
                    # CREATE INDEX gidx ON g (label1);
            
            insertionQuery = "INSERT INTO g select * from " + sys.argv[2]
            
        self.cur.execute(insertionQuery)


    def closeQuery(self):
        print("Closing Query.")
        resultsFile = open("results.txt", "w")
        node = self.root
        workingTuple = (node.subexp.value,)
            
        if (self.isFour):        
            self.cur.execute("SELECT DISTINCT * FROM g WHERE label1 = %s", workingTuple)
            #outputQuery = '''select * from resultoutput4ary(%s)'''
            #self.cur.execute(outputQuery, workingTuple)
        else:
            self.cur.execute("SELECT DISTINCT compstart, compend FROM g WHERE label1 = %s", workingTuple)
            
        counter = 0
        for results in self.cur:
            resultsFile.write(str(results) + "\n")
            counter = counter + 1
        
        resultsFile.close()
        self.cur.execute('''DROP TABLE g''')
        self.conn.close()
        self.cur.close()
        self.t1 = time.time()
        print "Time to process the query: " + str(self.t1-self.t0)
    
        print "Number of results: " + str(counter)


class RPQEngine:
    
    def __init__(self, isFour):
        self.isFour = isFour

    def run(self, query):
        char_stream = ANTLRStringStream(query)
        lexer = RegExpLexer(char_stream)
        tokens = CommonTokenStream(lexer)
        parser = RegExpParser(tokens)
        r = parser.exp()
        # this is the root of the AST
        root = r.tree
        print root.toStringTree()
        walker = TreeWalker(root, self.isFour)
        walker.postorder(root)
        walker.doWalk(root)
        print root.subexp.value


if __name__ == '__main__':
    if (len(sys.argv) < 4):
        print("Usage: python RPQEngine.py (query) (tablename) (config file) (-4 for returning paths)")
    else:    
        if (len(sys.argv) == 5 and sys.argv[4] == "-4"):
            calc = RPQEngine(True)
        else:     
            calc = RPQEngine(False)
        calc.run(sys.argv[1])
        
        

File: RPQEngine/src/main/java/org/dataone/daks/rpq/RPQEngine.java
package org.dataone.daks.rpq;

import java.util.Hashtable;
import org.dataone.daks.regexpparser.ParserExecutor;
import org.antlr.runtime.tree.CommonTree;


public class RPQEngine {
	
	private boolean isFour;
	private RPQDBDAO dao;
	private String tableName;
	
	
	public RPQEngine() {
		this.isFour = false;
		this.dao = RPQDBDAO.getInstance();
	}
	
	
	public boolean isFour() {
		return this.isFour;
	}
	
	
	public void setIsFour(boolean value) {
		this.isFour = value;
	}
	
	
	public void initDAO(String configFile) {
		this.dao.initFromConfigFile(configFile);
	}
	
	
	public static void main( String[] args ) {
		if( args.length < 3) {
			System.out.println( "Usage: java RPQEngine (query) (tablename) (config file) (-4 for returning paths)" );
			System.exit(0);
		}
		RPQEngine engine = new RPQEngine();
		engine.initDAO(args[2]);
		if ( (args.length == 4) && (args[3].equals("-4")) )
			engine.setIsFour(true);
		engine.tableName = args[1];
		engine.executeQuery(args[0]);
	}
    
	
	public void initQuery() {
		if(!this.isFour)
			this.dao.createBinaryTable(this.tableName);
		else
			this.dao.createQuaternaryTable(this.tableName);
	}
	
	
	public void executeQuery(String query) {
		if(this.tableName == null) {
			System.out.println("Error: the table name has not been specified");
			System.exit(0);
		}
		long t0 = System.currentTimeMillis();
		this.initQuery();
		ParserExecutor parserExec = new ParserExecutor();
		CommonTree root = parserExec.createParseTree(query);
		System.out.println(root.toStringTree());
		Hashtable<CommonTree, String> subExpHT = new Hashtable<CommonTree, String>();
		this.postorder(root, subExpHT);
		System.out.println("Evaluating: " + subExpHT.get(root));
		this.walkTree(root, subExpHT);
		this.closeQuery(root, subExpHT);
		long t1 = System.currentTimeMillis();
        System.out.println("Time to process the query: " + ((t1-t0)/1000.0) + " seconds.");
	}
	
	
	private void postorder(CommonTree tree, Hashtable<CommonTree, String> subExpHT) {
        for(int i = 0; i < tree.getChildCount(); i++ ) {
        	CommonTree child = (CommonTree)tree.getChild(i);
        	postorder(child, subExpHT);
        }
        String subExprStr = "";
        for(int i = 0; i < tree.getChildCount(); i++ ) {
        	CommonTree child = (CommonTree)tree.getChild(i);
        	subExprStr += subExpHT.get(child) + " ";
        }
        subExprStr += tree.getText() + " ";
        subExpHT.put(tree, subExprStr.trim());
	}
	
	
	private void walkTree(CommonTree tree, Hashtable<CommonTree, String> subExpHT) {
		for(int i = 0; i < tree.getChildCount(); i++ ) {
        	CommonTree child = (CommonTree)tree.getChild(i);
        	walkTree(child, subExpHT);
        }
		if(tree.getText().equals("|"))
			this.outputOr(tree, subExpHT);
		else if(tree.getText().equals("."))
			this.outputConc(tree, subExpHT);
		else if(tree.getText().equals("*"))
			this.outputStar(tree, subExpHT);
		else if(tree.getText().equals("+"))
			this.outputPlus(tree, subExpHT);
		else if(tree.getText().equals("-"))
			this.outputMinus(tree, subExpHT);
		else if(tree.getText().equals("?"))
			this.outputOptional(tree, subExpHT);
	}
	
	
	private void outputOr(CommonTree node, Hashtable<CommonTree, String> subExpHT) {
		String orExpr = subExpHT.get(node);
		String e1 = subExpHT.get(node.getChild(0));
		String e2 = subExpHT.get(node.getChild(1));
		String orQuery = null;
		if(this.isFour)
			orQuery = "INSERT INTO g SELECT compstart, '" + orExpr + "', compend, basestart, label2, baseend " +
						"FROM g WHERE label1 = '" + e1 + "' OR label1 = '" + e2 + "'";
		else
			orQuery = "INSERT INTO g SELECT compstart, '" + orExpr + "', compend " +
					"FROM g WHERE label1 = '" + e1 + "' OR label1 = '" + e2 + "'";
		this.dao.executeStatement(orQuery);
	}
	
	
	private void outputConc(CommonTree node, Hashtable<CommonTree, String> subExpHT) {
		String concExpr = subExpHT.get(node);
		String e1 = subExpHT.get(node.getChild(0));
		String e2 = subExpHT.get(node.getChild(1));
		String concQuery = null;
		if(this.isFour) {
			concQuery = "INSERT INTO g SELECT g1.compstart, '" + concExpr + "', g2.compend, g1.basestart, " + 
						"g1.label2, g1.baseend FROM g as g1, g as g2 " + 
						"WHERE g1.label1 = '" + e1 + "' and g1.compend = g2.compstart and  g2.label1 = '" + e2 + "'";
			this.dao.executeStatement(concQuery);
			concQuery = "INSERT INTO g SELECT g1.compstart, '" + concExpr + "', g2.compend, g2.basestart, " + 
					"g2.label2, g2.baseend FROM g as g1, g as g2 " + 
					"WHERE g1.label1 = '" + e1 + "' and g1.compend = g2.compstart and  g2.label1 = '" + e2 + "'";
			this.dao.executeStatement(concQuery);
		}
		else {
			concQuery = "INSERT INTO g SELECT g1.compstart, '" + concExpr + "', g2.compend " + 
					"FROM g as g1, g as g2 " + 
					"WHERE g1.label1 = '" + e1 + "' and g1.compend = g2.compstart and  g2.label1 = '" + e2 + "'";
			this.dao.executeStatement(concQuery);
		}
	}
	
	
	private void outputStar(CommonTree node, Hashtable<CommonTree, String> subExpHT) {
		String starExpr = subExpHT.get(node);
		String starSubExpr = subExpHT.get(node.getChild(0));
		String tcQuery = null;
		if(this.isFour) {
            tcQuery = "select * from tcwithrecursive4ary('" + starSubExpr + "', ' *' )";
            this.dao.executeStatement(tcQuery);
            String selfEdgeQueryLeft = "INSERT INTO g " +
                "select g.basestart, '" + starExpr + "', g.basestart, g.basestart, '" + starExpr + "', " + 
            	"g.basestart from g where g.label1 = g.label2";
            String selfEdgeQueryRight = "INSERT INTO g " +
                "select g.baseend, '" + starExpr + "', g.baseend, g.baseend, '" + starExpr + "', " + 
            	"g.baseend from g where g.label1 = g.label2";
            this.dao.executeStatement(selfEdgeQueryLeft);
            this.dao.executeStatement(selfEdgeQueryRight);
        }
        else {
        	tcQuery = "select * from tcwithrecursive('" + starSubExpr + "', ' *' )";
        	this.dao.executeStatement(tcQuery);
            String selfEdgeQuery = "INSERT INTO g " +
                "select distinct g.compstart, '" + starExpr + "', g.compstart from g " +
                "union " +
                "select distinct g.compend, '" + starExpr + "', g.compend from g";
            this.dao.executeStatement(selfEdgeQuery);
        }
	}
	
	
	private void outputPlus(CommonTree node, Hashtable<CommonTree, String> subExpHT) {
		String plusSubExpr = subExpHT.get(node.getChild(0));
		String tcQuery = null;
		if(this.isFour) {
            tcQuery = "select * from tcwithrecursive4ary('" + plusSubExpr + "', ' +' )";
            this.dao.executeStatement(tcQuery);
        }
        else {
        	tcQuery = "select * from tcwithrecursive('" + plusSubExpr + "', ' +' )";
        	this.dao.executeStatement(tcQuery);
        }
	}
	
	
	private void outputMinus(CommonTree node, Hashtable<CommonTree, String> subExpHT) {
		String minusExpr = subExpHT.get(node);
		String minusSubExpr = subExpHT.get(node.getChild(0));
		String minusQuery = null;
		if(this.isFour) {
            minusQuery = "INSERT INTO g SELECT compend, '" + minusExpr + "', compstart, basestart, label2, baseend" + 
            			"FROM g WHERE label1 = '" + minusSubExpr + "'";
            this.dao.executeStatement(minusQuery);
        }
        else {
        	minusQuery = "INSERT INTO g SELECT compend, '" + minusExpr + "', compstart " + 
        			"FROM g WHERE label1 = '" + minusSubExpr + "'";
        	this.dao.executeStatement(minusQuery);
        }
	}
	
	
	private void outputOptional(CommonTree node, Hashtable<CommonTree, String> subExpHT) {
		String optExpr = subExpHT.get(node);
		String optSubExpr = subExpHT.get(node.getChild(0));
		String selfEdgeQueryLeft = null;
		String selfEdgeQueryRight = null;
		String optQuery = null;
		if(this.isFour) {
			selfEdgeQueryLeft = "INSERT INTO g " + 
            		"select g.basestart, '" + optExpr + "', g.basestart, g.basestart, '" + optSubExpr + 
            		"', g.basestart from g where g.label1 = g.label2";
            this.dao.executeStatement(selfEdgeQueryLeft);
            selfEdgeQueryRight = "INSERT INTO g " + 
            		"select g.baseend, '" + optExpr + "', g.baseend, g.baseend, '" + optSubExpr + 
            		"', g.baseend from g where g.label1 = g.label2";
            this.dao.executeStatement(selfEdgeQueryRight);
            optQuery = "INSERT INTO g " + 
            		"SELECT compstart, '" + optExpr + "', compend, basestart, label2, baseend " + 
            		"FROM g WHERE label1 = '" + optSubExpr + "'";
            this.dao.executeStatement(optQuery);
        }
        else {
        	selfEdgeQueryLeft = "INSERT INTO g select g.compstart, '" + optExpr + "', g.compstart from g";
            selfEdgeQueryRight = "INSERT INTO g select g.compend, '" + optExpr + "', g.compend from g";
            this.dao.executeStatement(selfEdgeQueryLeft);
            this.dao.executeStatement(selfEdgeQueryRight);
            optQuery = "INSERT INTO g SELECT compstart, '" + optExpr + 
            		"', compend FROM g WHERE label1 = '" + optSubExpr + "'";
            this.dao.executeStatement(optQuery);
        }
	}
    
	
	public void closeQuery(CommonTree root, Hashtable<CommonTree, String> subExpHT) {
		String expr = subExpHT.get(root);
		this.dao.outputResults(expr, this.isFour);
	}
	
	
}


